packages/flutter/lib/src/gestures/resampler.dart

```diff
@@ -0,0 +1,428 @@
+        // early.
+        //
+        // When the input frequency is no greater than the sampling
+        // frequency, this early processing should guarantee that `up` and
+        // `remove` events are always re-sampled.
+        final Duration nextTimeStamp = _next?.timeStamp ?? const Duration();
+        assert(event.timeStamp >= nextTimeStamp);
+        if (event.timeStamp > nextTimeStamp) {
+          break;
+        }
+      }
+
+      final bool wasTracked = _isTracked;
+      final bool wasDown = _isDown;
+
+      // Update `_Pointer` state.
+      if (event is PointerDownEvent || event is PointerMoveEvent) {
+        _isDown = true;
+        _isTracked = true;
+      } else if (event is PointerUpEvent || event is PointerCancelEvent) {
+        _isDown = false;
+        _isTracked = true;
+      } else if (event is PointerAddedEvent || event is PointerHoverEvent) {
+        _isTracked = true;
+      } else if (event is PointerRemovedEvent) {
+        _isTracked = false;
+      }
+
+      // Position at `sampleTime`.
+      final Offset position = _positionAt(sampleTime);
+
+      // Initialize position if we are starting to track this pointer.
+      if (_isTracked && !wasTracked) {
+        _position = position;
+      }
+
+      // Current pointer identifier.
+      final int pointerIdentifier = event.pointer;
+
+      // Initialize pointer identifier for `move` events.
+      // Identifier is expected to be the same while `down`.
+      assert(!wasDown || _pointerIdentifier == pointerIdentifier);
+      _pointerIdentifier = pointerIdentifier;
+
+      // Skip `move` and `hover` events as they are automatically
+      // generated when the position has changed.
+      if (event is! PointerMoveEvent && event is! PointerHoverEvent) {
+        dequeuedEvents.add(_clone(event, position, position - _position,
+            pointerIdentifier, sampleTime));
+        _position = position;
+      }
+
+      _queuedEvents.removeFirst();
+    }
+
+    return dequeuedEvents;
+  }
+
+  List<PointerEvent> _samplePointerPositions(Duration sampleTime) {
+    final List<PointerEvent> sampledEvents = <PointerEvent>[];
+
+    // Position at `sampleTime`.
+    final Offset position = _positionAt(sampleTime);
+
+    // Add `move` or `hover` events if position changed.
+    final PointerEvent? next = _next;
+    if (position != _position && next != null) {
+      final Offset delta = position - _position;
+      final PointerEvent event = _isDown
+          ? _toMoveEvent(next, position, delta, _pointerIdentifier, sampleTime)
+          : _toHoverEvent(next, position, delta, sampleTime);
+      sampledEvents.add(event);
+      _position = position;
+    }
+
+    return sampledEvents;
+  }
+
+  /// Enqueue pointer `event` for resampling.
+  void addEvent(PointerEvent event) {
+    _queuedEvents.add(event);
+  }
+
+  /// Dequeue resampled pointer `event` for the specified `sampleTime`.
+  List<PointerEvent> sample(Duration sampleTime) {
```

Comment:
I see in the unit test this can return a list because there may be multiple events at the same time. Does it also makes sense to comment it here in the doc comment?

Reply:
Done.